### PR TITLE
3346 migrating to 4.2

### PR DESCRIPTION
**Pull request by baburdick**
Task 3346

* Strips out Foreigner
* Embraces Rails 4.2's foreign key support

Built on @MityaLiu's work toward Rails 4.2 migration.

### CODE REVIEW

**Comment by baburdick** [comment]
Migrations do not yet complete:
```
PgSaurus::IndexExistsError: The index, index_demography_cities_on_country_id, already exists.  Use :exclude_index => true when adding the foreign key.
~/Code/HornsAndHooves-pg_saurus/lib/pg_saurus/connection_adapters/postgresql_adapter/foreigner_methods.rb:95:in `add_foreign_key'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/migration.rb:665:in `block in method_missing'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/migration.rb:634:in `block in say_with_time'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/migration.rb:634:in `say_with_time'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/migration.rb:654:in `method_missing'
~/Code/HornsAndHooves-pg_saurus/spec/dummy/db/schema.rb:107:in `block in <top (required)>'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/schema.rb:41:in `instance_eval'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/schema.rb:41:in `define'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/schema.rb:61:in `define'
~/Code/HornsAndHooves-pg_saurus/spec/dummy/db/schema.rb:14:in `<top (required)>'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:268:in `load'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:268:in `block in load'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:268:in `load'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/tasks/database_tasks.rb:218:in `load_schema_for'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/railties/databases.rake:325:in `block (3 levels) in <top (required)>'
~/.rvm/gems/ruby-2.2.4@pg_saurus/gems/activerecord-4.2.6/lib/active_record/railties/databases.rake:314:in `block (3 levels) in <top (required)>'
~/.rvm/gems/ruby-2.2.4@pg_saurus/bin/ruby_executable_hooks:15:in `eval'
~/.rvm/gems/ruby-2.2.4@pg_saurus/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by baburdick** [comment]
Confirmed that specs pass! :-) 👍

File: spec/foreign_keys_spec.rb [hunk 1]
```diff
@@ -9,48 +9,19 @@
...
-    end
-
-    # AddForeignKeys migration
     #   add_foreign_key 'demography.citizens', 'users', :exclude_index => true
     it 'should not add an index on the foreign key when :exclude_index is true' do
       PgSaurus::Explorer.index_exists?('demography.citizens', :user_id).should == false
     end
 
     it 'should raise a PgSaurus::IndexExistsError when the index already exists' do
+      puts ActiveRecord::Base::connection.indexes('demography.citizens')
```
Comment (baburdick) [inline spec/foreign_keys_spec.rb]: @albertosaurus: diagnostic code?

Comment (albertosaurus) [inline spec/foreign_keys_spec.rb]: yes

File: lib/pg_saurus/connection_adapters/postgresql_adapter/foreigner_methods.rb [hunk 2]
```diff
@@ -1,221 +0,0 @@
...
-      #   So, in case of concurrently created index with foreign key only
-      #   foreign key will be created inside migration transaction and after
-      #   closing transaction queries for index creation will be send to database.
-      #   That's why I prevent here normal index creation in case of
-      #   `concurrent_index` option is given.
-      #   NOTE: Index creation after closing migration transaction could lead
-      #   to weird effects when transaction moves smoothly, but index
-      #   creation with error. In that case transaction will not be rolled back.
-      #   As it was closed before even index was attempted to create.
-      #   -- zekefast 2012-09-12
```
Comment (baburdick) [inline lib/pg_saurus/connection_adapters/postgresql_adapter/foreigner_methods.rb]: @albertosaurus: is zekefast's warning no longer germane?

Comment (albertosaurus) [inline lib/pg_saurus/connection_adapters/postgresql_adapter/foreigner_methods.rb]: No. What we were doing before was silently not creating the index. Not what I want to do going forward.

File: README.markdown [hunk 3]
```diff
@@ -1,27 +1,43 @@
...
-PgSaurus was tested with Rails 4.0.x, Ruby 2.1.2.
+PgSaurus v3 was tested with Rails 4.2, Ruby 2.2.4. For Rails 4.1, use PgSaurus v2.5+.
+
+NOTE: JRuby is not supported. The current ActiveRecord JDBC adapter has its own Rails4-compatible
+method named "create_schema" which conflicts with this gem.
 
-NOTE: JRuby is not yet supported. The current ActiveRecord JDBC
-adapter has its own Rails4-compatible method named "create_schema" which
-conflicts with this gem.
+NOTE: PgSaurus does not support Rails 3.
```
Comment (greyblake) [inline README.markdown]: You can put reference to pg_power here, for Rails 3

File: spec/dummy/config/database.yml [hunk 4]
```diff
@@ -4,9 +4,9 @@ development: &default
   encoding: unicode
   database: pg_saurus_dummy_development
   pool: 5
-  username: postgres
-  password: secret
-  host: localhost
+  #username: postgres
+  #password: secret
+  #host: localhost
```
Comment (greyblake) [inline spec/dummy/config/database.yml]: Maybe this  file should renamed to `database.yml.example` and `spec/dummy/config/database.yml` should be put to `.gitignore`

**Comment by greyblake** [comment]
Nice to see you still use this library, guys! Long live PgSaurus!

File: spec/dummy/config/database.yml (hunk 4 above)
Comment (albertosaurus) [inline spec/dummy/config/database.yml]: Not a bad notion, actually. Maybe after we pull the trigger on this?

File: README.markdown (hunk 3 above)
Comment (albertosaurus) [inline README.markdown]: pg_power is now completely unsupported.

**Pull request merged**